File: VisualStudio/AI_homework_singleneuron/AI_homework_singleneuron/singleneuron.c
/*
코드 제목 : 인공신경망 알고리즘을 활용하여 1개의 노드 학습
코드 세부 내용 : Input값이 1일때 Output 값이 5가 되도록 1개의 노드를 교사학습
*/

#include <stdio.h>
#include <math.h>

#define INPUT 1
#define BIAS 1
#define TARGET_OUTPUT 5

int main() {
	double weight = 1, alpha = 0.1, sigma, y;
	int i = 1;

	printf("-----------------------------------------------------------------------------------------------------------------------------------------------------\n");
	printf(" Is training Success?            output              target_output        Training Rate               Weight                       Error             \n");

	// Runnig 100 Times for each Training Rate
	while (alpha <= 0.9) {
		printf("-----------------------------------------------------------------------------------------------------------------------------------------------------\n");
		while (i <= 100) {

			// SIGMA function
			sigma = INPUT * weight + BIAS;

			// Activation function
			y = sigma;

			if (y == TARGET_OUTPUT) {
				printf("          Yes            %.15e             %d                %lf           %.15e       %.15e\n", y, TARGET_OUTPUT, alpha, weight, pow(TARGET_OUTPUT - y, 2) / 2);
			}
			else {
				printf("          No             %.15e             %d                %lf           %.15e       %.15e\n", y, TARGET_OUTPUT, alpha, weight, pow(TARGET_OUTPUT - y, 2) / 2);
			}

			// Training weigiht
			weight = weight - alpha * (y - TARGET_OUTPUT);
			i++;
		}

		alpha += 0.1;
		i = 1;
		weight = 1;
	}
	printf("-----------------------------------------------------------------------------------------------------------------------------------------------------\n");
	printf(" Is training Success?            output              target_output        Training Rate               Weight                       Error             \n");
	printf("-----------------------------------------------------------------------------------------------------------------------------------------------------\n");
	return 0;
}

File: Pycharm/AI/singleneuron.py
'''
제목 : SingleNeuron.py
주제 : 단일 뉴런 weight 및 bias 학습 및 그래프
'''

import matplotlib.pyplot as plt

training_rate_arr = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]

input, target = 1, 5

for training_rate in training_rate_arr :
    weight, bias = 1., 1.
    weight_arr = []
    error_arr = []
    print("*************** Trainig rate is %f ***************" %(training_rate))
    for i in range (1,101) :
        sigma = weight*input+bias
        output = sigma
        error = (target-output)**2*0.5
        if output==target :
            print("Success   %d times   output is %.15e   weight is %.15e   bias is %.15e   error rate is %.15e" %(i, output, weight, bias, error))
        else :
            print("Fail      %d times   output is %.15e   weight is %.15e   bias is %.15e   error rate is %.15e" %(i, output, weight, bias, error))

        weight_arr.append(weight)
        error_arr.append(error)

        weight = weight - training_rate*(output-target)
        bias = bias - training_rate*(output-target)

    plt.title("Training rate is %f" % training_rate)
    i = range(1, 101)
    plt.plot(i, weight_arr, color='blue', linestyle = '-', label='weight')
    plt.plot(i, error_arr, color='red', linestyle = '-', label = 'error rate')
    plt.legend(loc='upper right', frameon = False)
    plt.show()

